src/braket/experimental/autoqasm/program/program.py

```diff
@@ -89,6 +90,129 @@ class ProgramMode(Enum):
+
+        Args:
+            ir_type (IRType): The IRType to use for displaying the program.
+                Defaults to IRType.OPENQASM.
+        """
+
+
+class MainProgram(Program):
+    def __init__(self, program_generator: Callable[[None], Program]):
+        self._program_generator = program_generator
+        self._program = None
+
+    def _ensure_initialized(self) -> None:
+        if not self._program:
+            self._program = self._program_generator()
+
+    def to_ir(
+        self,
+        ir_type: IRType = IRType.OPENQASM,
+        serialization_properties: SerializationProperties = OpenQASMSerializationProperties(),
+    ) -> str:
+        """Serializes the program into an intermediate representation.
+
+        Args:
+            ir_type (IRType): The IRType to use for converting the program to its
+                IR representation. Defaults to IRType.OPENQASM.
+            serialization_properties (SerializationProperties): IR serialization configuration.
+                Default to OpenQASMSerializationProperties().
+
+        Raises:
+            ValueError: If the supplied `ir_type` is not supported.
+
+        Returns:
+            str: A representation of the program in the `ir_type` format.
+        """
+        self._ensure_initialized()
+        return self._program.to_ir(ir_type, serialization_properties)
+
+    def with_calibrations(self, gate_calibrations: Union[Callable, list[Callable]]) -> Program:
+        """Add the gate calibrations to the program. The calibration added program is returned
+        as a new object. The original program is not modified.
+
+        Args:
+            gate_calibrations (Union[Callable, list[Callable]]): The gate calibrations to add to
+                the main program. Calibration are passed as callable without evaluation.
+
+        Returns:
+            Program: The program with gate calibrations added.
+        """
+        self._ensure_initialized()
+        return self._program.with_calibrations(gate_calibrations)
+
+    def make_bound_program(self, param_values: dict[str, float], strict: bool = False) -> Program:
+        """Binds FreeParameters based upon their name and values passed in.
+
+        Args:
+            param_values (dict[str, float]): A mapping of FreeParameter names
+                to a value to assign to them.
+            strict (bool): If True, raises a ParameterNotFoundError if any of the FreeParameters
+                in param_values do not appear in the program. False by default.
+
+        Raises:
+            ParameterNotFoundError: If a parameter name is given which does not appear in
+                the program.
+
+        Returns:
+            Program: Returns a program with all present parameters fixed to their respective
+            values.
+        """
+        self._ensure_initialized()
+        return self._program.make_bound_program(param_values, strict)
+
+    def display(self, ir_type: IRType = IRType.OPENQASM) -> None:
```

Comment:
This is `print(prog.to_ir())`? We might want to reserve `display` for drawing the program, or some other improved visualization.

Reply:
Today this is the "pretty print" API that uses `pygments`. But I agree with you. I am also thinking that `display` and similar functionality belong in some separate visualization module rather than being part of the `Program` API.